### PR TITLE
Caesar encryption

### LINKED ISSUE #26: Caesar Encryption

### PR DESCRIPTION
**Pull request by CaesarGhazi**
---
name: Caesar Ghazi
about: caesar_encryption
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Comment by Nada-saad635** [comment]
Nada is reviewing this

File: solutions/caesar_encryption.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+A module for encrypting using caesar's cypher.
+
+Module contents:
+    - caesar_encryption: generates an encrypted string.
+
```
Comment (Nada-saad635) [inline solutions/caesar_encryption.py]: i think here you have to put more explanation about the model and how it works

File: solutions/caesar_encryption.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        text (str): The input text to encrypt.
+        shift (int): The number of positions to shift each letter.
+
+    Returns:
+        str: The encrypted text.
+
+    Raises:
+        AssertionError: if Text is not a string.
+        AssertionError: if Shift is not an integer.
+
```
Comment (Nada-saad635) [inline solutions/caesar_encryption.py:34]: here you are missing smth like Other assumptions (eg. if it's a number, what's the expected range?)
Ensures the shift is within 0-25

File: solutions/caesar_encryption.py (hunk 1 above)
Comment (CaesarGhazi) [reply solutions/caesar_encryption.py]: thankyou so much for taking the time to review this Nada, i really appreciate this.
i will add ir right now.

File: solutions/caesar_encryption.py (hunk 2 above)
Comment (CaesarGhazi) [reply solutions/caesar_encryption.py:34]: i'm sorry but i wanna make sure i get this correctly, do you mean i should make the shift is within a specific range or what? 
because right now i just made the mod 26 so it wraps around if it exceeds the alphabet range (0-25).

**Comment by omniaNS** [comment]
Nice work on both the function and the tests! Code is clear and easy to understand. I don’t have anything to add, everything looks solid👍

**Review by omniaNS — COMMENTED** [review]
Nice work on both the function and the tests! Code is clear and easy to understand. I don’t have anything to add, everything looks solid👍

File: solutions/caesar_encryption.py (hunk 2 above)
Comment (Nada-saad635) [reply solutions/caesar_encryption.py:34]: yea  i know i see that but its better to right it as comment and explain it

**Comment by Nada-saad635** [comment]
otherwise the code and tests are good nice work

**Comment by CaesarGhazi** [comment]
thank you so much both for your effort and for taking the time to review this challenge, i truly appreciate it .

File: solutions/caesar_encryption.py (hunk 2 above)
Comment (CaesarGhazi) [reply solutions/caesar_encryption.py:34]: i added your suggestion right now, thankyooou sm 😊

File: solutions/tests/test_caesar_encryption.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    # Defensive Tests
+    def test_invalid_text_type(self):
+        """it should raise an AssertionError if the text parameter is not a string."""
+        with self.assertRaises(AssertionError):
+            caesar_encryption(123, 3)
+
+    def test_invalid_shift_type(self):
+        """it should raise an AssertionError if the shift parameter is not an integer."""
+        with self.assertRaises(AssertionError):
+            caesar_encryption("Hello", "three")
```
Comment (Nada-saad635) [inline solutions/tests/test_caesar_encryption.py:76]:
```suggestion
            if __name__ == "__main__":
                  unittest.main()
```

Comment (Nada-saad635) [reply solutions/tests/test_caesar_encryption.py:76]: add  these lines at the end

Comment (CaesarGhazi) [reply solutions/tests/test_caesar_encryption.py:76]: I've always wanted to ask this, idk why I didn't.
But can I ask what do those lines exactly do 🫣😂😂😂?

Comment (CaesarGhazi) [reply solutions/tests/test_caesar_encryption.py:76]: But don't worry I'll add them right now

Comment (Nada-saad635) [reply solutions/tests/test_caesar_encryption.py:76]: To ensure that the code inside this block is only executed when the script is run directly, not when it is imported.

**Review by Nada-saad635 — APPROVED** [review]

**Comment by Nada-saad635** [comment]
nice work ! ready to merege
